### PR TITLE
Add mark spam button

### PR DESCRIPTION
**Pull request by ldwiyono**
Adds mark_spam_confirm_modal in core folder to be implemented for Events, Jobs, Codebases. Essentially resolves comses/planning/issues/255, however improvements in UX and further functionalities are still possible.

Current appearance on desktop for events:
image

### CODE REVIEW

File: django/core/mixins.py [hunk 1]
```diff
@@ -268,6 +270,32 @@ def _validate_content_object(self, instance):
                     f"instance {instance} does not have a {field} attribute"
                 )
 
+    @action(detail=True, methods=["post"])
+    def mark_spam(self, request, **kwargs):
+        # consider adding notes field to the request
```
Comment (sgfost) [inline django/core/mixins.py:275]: can remove

File: django/core/mixins.py [hunk 2]
```diff
@@ -268,6 +270,32 @@ def _validate_content_object(self, instance):
...
+                "notes": notes,
+            },
+        )
+        if not created:
+            spam_moderation.status = SpamModeration.Status.SPAM
+            spam_moderation.detection_method = "manual"
+            spam_moderation.reviewer = request.user
+            spam_moderation.notes = notes
+            spam_moderation.save()
+        # return Response({"message": "Successfully marked as spam!"}, status=status.HTTP_200_OK)
```
Comment (sgfost) [inline django/core/mixins.py:296]: this too

File: django/core/mixins.py [hunk 3]
```diff
@@ -268,6 +270,32 @@ def _validate_content_object(self, instance):
                     f"instance {instance} does not have a {field} attribute"
                 )
 
+    @action(detail=True, methods=["post"])
```
Comment (sgfost) [inline django/core/mixins.py:273]: the action decorator can take in `permission_classes` to override the viewset level permissions https://www.django-rest-framework.org/api-guide/viewsets/

File: django/core/jinja2/core/events/retrieve.jinja [hunk 4]
```diff
@@ -75,6 +75,7 @@
         <a href="{{ url('core:event-edit', pk=id) }}">
             <div class="btn btn-primary my-1 w-100">Edit</div>
         </a>
+        {{ mark_spam_confirm_modal("Mark spam", title, url("core:event-mark-spam", pk=id), csrf_input) }}
```
Comment (sgfost) [inline django/core/jinja2/core/events/retrieve.jinja:78]: should update the conditional rendering of the mark spam buttons to only show when the content object `is_marked_spam` is `False` and that the viewing user has permission

**Review by sgfost — CHANGES_REQUESTED** [review]
I just realized there are no permissions required to use the `mark_spam` action so anyone can mark anything as spam (though the button is only shown to users with edit perms for the object, so admins and submitters)

The action should be locked down and the button should only be rendered when the user has the same permission. Not 100% sure what that should be, however.

@alee What do you think the access here should be: admins, wagtail moderators, editors, or a granular permission?

Otherwise it looks good, just a couple comments to get rid of

**Comment by sgfost** [comment]
I think this would resolve comses/planning#201 which appears to talk about the `Moderators` group

**Comment by ldwiyono** [comment]
I'm ready to make these changes whenever necessary.  The conditional could be something like `{% if has_change_perms and has_moderator_access %}`?

**Comment by sgfost** [comment]
change perms aren't necessary. The condition on the template should be something like `IF NOT marked spam AND is a moderator`

I'd just go with the Moderator group for now and we can change it if needed. The permission class to assign to `mark_spam` should be straightforward, just need to make a custom DRF permission class (https://www.django-rest-framework.org/api-guide/permissions/#custom-permissions) that checks if the `request.user` is a Moderator https://stackoverflow.com/a/20110261

**Comment by ldwiyono** [comment]
Hi, I added the new permission class `MarkSpamPermissions` and locked down both the `mark_spam` action and the conditional rendering of the `mark_spam` button. I used the `is_superuser` attribute of the user for the permissions class; if this differs from "moderator" in significant ways, please let me know as I could not easily find a test for that group.

**Comment by sgfost** [comment]
`is_superuser` is a field on the django user that makes `has_perm()` always return true, the idea being that it's one thing to set to give a user the ability to do anything on the site. The `Moderator` role refers to a `Group` which is provided by wagtail and has slightly less access than a wagtail admin.

You'll just need to check whether the user is a part of the `"Moderators"` group https://stackoverflow.com/questions/4789021/in-django-how-do-i-check-if-a-user-is-in-a-certain-group. Should probably add moderators to the `ComsesGroups` enum if we are going to start using it explicitly for stuff

**Comment by sgfost** [comment]
also, no need to keep amending and force pushing the same commit. The initial rebase was just to take care of the test commits

**Comment by ldwiyono** [comment]
Sorry about force pushing the commits, I'll keep that in mind for the future. I'm currently trying this:

```
class MarkSpamPermissions(permissions.BasePermission):
    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        
        return request.user.groups.filter(name='Moderator').exists()

```
but when I'm logged in as admin_user, it denies me permission to call the ```mark_spam``` action.
image

**Comment by sgfost** [comment]
the `Admins` group (also from wagtail) is different from the `is_superuser` and `is_staff` flags on the user object. The admin_user must not have any of these set is and relying on being a superuser for access to everything. I think the solution would be to return true if `request.user.is_superuser` in the permission class before checking for group membership

**Comment by ldwiyono** [comment]
Okay, thanks. I've incorporated those changes into the newest commit; please let me know if this works for you.

**Review by sgfost — APPROVED** [review]

**Review by alee — CHANGES_REQUESTED** [review]
good work, thanks @ldwiyono and @sgfost !

some minor suggestions included inline, I think things would be cleaner if we pushed the `user.groups.filter(name=...).exists()` logic into `ComsesGroups.is_moderator(user)`

I've deployed for further testing on staging before merging. DM me on slack if you need access to a Moderator user account there, create a new user account if you want and I can add you to the group.

One thing immediately of note: I don't seem to be able to mark older events as spam (but ironically I can mark my own submissions as spam)

image

File: django/core/view_helpers.py [hunk 5]
```diff
@@ -79,4 +79,8 @@ def add_change_delete_perms(instance, data, user):
         "{}.delete_{}".format(instance._meta.app_label, instance._meta.model_name),
         instance,
     )
+    data["can_moderate"] = (
+        user.is_superuser
+        or user.groups.filter(name=ComsesGroups.MODERATOR.value).exists()
```
[the remark this answers is not in the document]
Comment (sgfost) [reply django/core/view_helpers.py:84]: oh right, should we just replace and migrate the Reviewer group in #748?

[the remark this answers is not in the document]
Comment (alee) [reply django/core/view_helpers.py:84]: Yep,  probably good to clean that up - somehow there are 337 members in the Reviewers group on prod

[the remark this answers is not in the document]
Comment (ldwiyono) [reply django/core/view_helpers.py]: I put a single (non-static) method `is_group_member` which can be called on all instances of ComsesGroups (e.g. `ComsesGroups.ADMIN.is_group_member(user)`. I'm not sure if this clashes with Karthik's work on the PeerReviewer object.

**Comment by ldwiyono** [comment]
I think anyone with moderator perms should now be able to access the "Mark spam" button. The `is_group_member` method of ComsesGroups that I added is not static and may need adjustment.

File: django/core/models.py [hunk 6]
```diff
@@ -57,7 +57,14 @@ def get_group(self):
         if _group is None:
             _group = self.group = Group.objects.get(name=self.value)
         return _group
-
+    
+    # can be called like #ComsesGroups.ADMIN.is_group_member(user)
+    def is_group_member(self, user):
+        for member in self.users():
+            if member.username == user.username:
+                return True
```
Comment (sgfost) [inline django/core/models.py:65]: IMO the instance method approach (as opposed to multiple static methods) is fine, though this implementation is less efficient than the direct query (`user.groups.filter(name=self.value).exists()`) that its meant to act as a convenience method for

Comment (alee) [reply django/core/models.py:65]: Yep, also @ldwiyono make sure your git commits follow the conventional commits guidelines:

https://www.conventionalcommits.org/en/v1.0.0/#summary

Comment (ldwiyono) [reply django/core/models.py:65]: Yes, sorry, that is inefficient. I will try and push the more efficient `return self.users().filter(username=user.username).exists()`.

Comment (alee) [reply django/core/models.py]: I stuck with the static methods approach for now, just pushed a commit with some adjustments to the boolean logic in view_helpers.py as well

**Review by alee — APPROVED** [review]
LGTM, thanks @ldwiyono and @sgfost !

**Pull request merged**